[docs] Improve pagination documentation page

flaviendelangle opened the pull request:
[Doc preview](https://deploy-preview-3424--material-ui-x.netlify.app/components/data-grid/pagination/)

- [x] Add support for pageSize change on the service-side pagination demo

- [x] Add dynamic container height on the `autoHeight` demo to show how the grid reacts

- [x] New demo for controlled page size

- [x] Remove the `BasicPaginationGrid` and `200PaginationGrid` demo, I don't think they bring any value

- [x] Remove the `ApiRefPaginationGrid` (was not used since #3266)

- [x] Reorganize the paragraphs to have a better hierarchy

Before:
![image](https://user-images.githubusercontent.com/3309670/146153359-83421393-5cd9-492f-8d84-3e14bec941c3.png)

After:
![image](https://user-images.githubusercontent.com/3309670/146153405-3ac19b6f-eb36-4243-a347-7d9243c4f8a1.png)

alexfauquette commented on docs/src/pages/components/data-grid/pagination/pagination.md:
```diff
@@ -4,65 +4,75 @@ title: Data Grid - Pagination
+### Cursor implementation
 
-You can customize the rendering of the pagination in the footer following [the component section](/components/data-grid/components/#pagination) of the documentation.
+You can also create a cursor-based pagination.
```
"create" makes me feel it is a new feature, I did not get it was the server specificity

```suggestion
You can also handle servers with cursor-based pagination.
```

flaviendelangle replied:
Done

alexfauquette commented on docs/src/pages/components/data-grid/pagination/PageSizeAuto.js:
```diff
@@ -0,0 +1,45 @@
+        </Select>
+      </FormControl>
+      <div style={{ height, width: '100%' }}>
+        <DataGrid autoPageSize pagination {...data} />
```
`pagination` prop is in each example, even if it is useless because they all use `DataGrid` and not `DataGridPro`. Is it to remind `DataGridPro` users they need this prop?

flaviendelangle replied:
Yes, it was @oliviertassinari proposal.
I think @m4theushw agree with me that we should drop it.

alexfauquette replied:
Moreover, you insisted on the different default behavior between the two plans at the beginning. Could use a warning box to be sure devs do not miss it.

> The default pagination behavior depends on your plan.
> - On the `DataGrid`, pagination is enabled by default and can't be disabled
> - On the `DataGridPro`, pagination is disabled by default, use the `pagination` prop to enable it

flaviendelangle replied:
Done

oliviertassinari replied:
> Is it to remind DataGridPro users they need this prop?

It's to make sure that developers can replace `DataGrid` with `DataGridPro` without surprises.

alexfauquette reviewed (commented):
In the example "Page size options" props `pageSize` and `onPageSizeChange` are not usefull since you created a dedicated section for controlled pageSize.

I think, the order
- page size options
- automatic page size
- controlled paged size

would make more sense, because the first two are features that can be useful on their own. The last one is closely related to server-side filtering.

flaviendelangle commented:
> In the example "Page size options" props pageSize and onPageSizeChange are not usefull since you created a dedicated section for controlled pageSize.

Removed

> I think, the order [...] would make more sense

Updated, it's a lot better in this order :+1:

flaviendelangle commented on docs/src/pages/components/data-grid/pagination/pagination.md:
```diff
@@ -4,65 +4,75 @@ title: Data Grid - Pagination
 # Data Grid - Pagination
 
-<p class="description">Through pagination, a segment of data can be viewed from the assigned data source.</p>
+<p class="description">Easily pagination your rows and only fetch what you need.</p>
```
Yes :grimacing:

oliviertassinari replied:
I believe it's usually a technical writing best practice to avoid using terms like "easy". I saw this mentioned in https://youtu.be/q_LlUEIJnA0?t=91.

flaviendelangle commented on docs/src/pages/components/data-grid/pagination/ServerPaginationGrid.tsx:
```diff
@@ -1,57 +1,78 @@
       <DataGrid
-        rows={rows}
-        columns={data.columns}
+        columns={COLUMNS}
```
Oh I forgot to remove that one, I wanted to make the lazy loading cleaner but it requires to much work for this PR.

alexfauquette reviewed (approved):
I let you run formatting scripts, but for me it is good. Let see if @oliviertassinari really want to keep the `pagination` prop

oliviertassinari commented on docs/src/pages/components/data-grid/pagination/pagination.md:
```diff
@@ -4,65 +4,75 @@ title: Data Grid - Pagination
+You can also handle servers with cursor-based pagination.
+To do so, you just have to keep track of the next cursor associated with each page you fetched.
 
-## Paginate > 100 rows [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)
```
Why remove this demo? We broke these two anchor links

<img width="604" alt="Screenshot 2021-12-31 at 19 38 26" src="https://user-images.githubusercontent.com/3165635/147835784-35ea5526-2617-42b6-9268-03feaae0fe67.png">

https://mui.com/components/data-grid/getting-started/#feature-comparison

<img width="483" alt="Screenshot 2021-12-31 at 19 39 06" src="https://user-images.githubusercontent.com/3165635/147835792-71eb8195-afa7-4e25-861f-4d0e0d98715b.png">

https://mui.com/pricing/

flaviendelangle replied:
... I should have checked if there was links

Because I don't see the point of having this kind of demo.
Users do not learn anything from then compared to a line "only the pro version can paginate with more than 100 rows"
We now have long doc pages and I think we should focus on demos that demonstrate something.

oliviertassinari replied:
> I should have looked for links

@flaviendelangle I think that it's very hard to do in practice. I bet that there are more instances of broken anchor links broken in mui.com that we are not aware of. I wish the ahrefs SEO crawler was also reporting this. I have created https://ahrefs.canny.io/site-audit/p/html-anchor-link.

> Users do not learn anything from then compared to a line "only the pro version can paginate with more than 100 rows"

I might not agree, but fair enough.

How about we only add the h2 back (without the demo)?
